client/device/cloudftd/create.go

```diff
@@ -0,0 +1,158 @@
+package cloudftd
+
+import (
+	"context"
+	"fmt"
+	"github.com/CiscoDevnet/terraform-provider-cdo/go-client/device"
+	"github.com/CiscoDevnet/terraform-provider-cdo/go-client/device/cloudfmc"
+	"github.com/CiscoDevnet/terraform-provider-cdo/go-client/device/cloudfmc/fmcplatform"
+	"github.com/CiscoDevnet/terraform-provider-cdo/go-client/internal/cdo"
+	"github.com/CiscoDevnet/terraform-provider-cdo/go-client/internal/http"
+	"github.com/CiscoDevnet/terraform-provider-cdo/go-client/internal/retry"
+	"github.com/CiscoDevnet/terraform-provider-cdo/go-client/internal/url"
+	"github.com/CiscoDevnet/terraform-provider-cdo/go-client/model/devicetype"
+	"github.com/CiscoDevnet/terraform-provider-cdo/go-client/model/ftd/license"
+	"github.com/CiscoDevnet/terraform-provider-cdo/go-client/model/ftd/tier"
+)
+
+type CreateInput struct {
+	Name             string
+	AccessPolicyName string
+	PerformanceTier  *tier.Type // ignored if it is physical device
+	Virtual          bool
+	Licenses         []license.Type
+}
+
+type CreateOutput struct {
+	Uid      string    `json:"uid"`
+	Name     string    `json:"name"`
+	Metadata *Metadata `json:"metadata"`
+}
+
+func NewCreateInput(
+	name string,
+	accessPolicyName string,
+	performanceTier *tier.Type,
+	virtual bool,
+	licenses []license.Type,
+) CreateInput {
+	return CreateInput{
+		Name:             name,
+		AccessPolicyName: accessPolicyName,
+		PerformanceTier:  performanceTier,
+		Virtual:          virtual,
+		Licenses:         licenses,
+	}
+}
+
+type createRequestBody struct {
+	FmcId      string          `json:"associatedDeviceUid"`
+	DeviceType devicetype.Type `json:"deviceType"`
+	Metadata   *Metadata       `json:"metadata"`
+	Name       string          `json:"name"`
+	State      string          `json:"state"` // TODO: use queueTriggerState?
+	Type       string          `json:"type"`
+	Model      bool            `json:"model"`
+}
+
+func Create(ctx context.Context, client http.Client, createInp CreateInput) (*CreateOutput, error) {
+
+	client.Logger.Println("creating cloud ftd")
+
+	// 1. read Cloud FMC
+	fmcRes, err := cloudfmc.Read(ctx, client, cloudfmc.NewReadInput())
+	if err != nil {
+		return nil, err
+	}
+
+	// 2. get FMC domain uid by reading Cloud FMC domain info
+	readFmcDomainRes, err := fmcplatform.ReadFmcDomainInfo(ctx, client, fmcplatform.NewReadDomainInfo(fmcRes.Host))
+	if err != nil {
+		return nil, err
+	}
+	if len(readFmcDomainRes.Items) == 0 {
+		return nil, fmt.Errorf("fmc domain info not found")
+	}
+
+	// 3. get access policies using Cloud FMC domain uid
+	accessPoliciesRes, err := cloudfmc.ReadAccessPolicies(
+		ctx,
+		client,
+		cloudfmc.NewReadAccessPoliciesInput(fmcRes.Host, readFmcDomainRes.Items[0].Uuid, 1000), // 1000 is what CDO UI uses
+	)
+	if err != nil {
+		return nil, err
+	}
+	selectedPolicy, ok := accessPoliciesRes.Find(createInp.AccessPolicyName)
```

Comment:
File this as a separate story if you have to, but I envisaged that we should create a data source to retrieve FMC access policies, and allow the access policy ID to be passed through to the FTD during onboarding that way.

Reply:
Yea I will leave that as separate story, they can pass the name through as well